applications/DigitalTwinApplication/custom_sensors/strain_sensor.cpp

```diff
@@ -0,0 +1,356 @@
+        "value"          : 0.0,
+        "location"       : [0.0, 0.0, 0.0],
+        "strain_type"    : "strain_xx",
+        "strain_variable": "SHELL_STRAIN",
+        "weight"         : 0.0
+    })" );
+    parameters["name"].SetString(this->GetName());
+    parameters["value"].SetDouble(this->GetSensorValue());
+    parameters["location"].SetVector(this->GetLocation());
+    parameters["weight"].SetDouble(this->GetWeight());
+
+    switch (mStrainType) {
+        case StrainType::STRAIN_XX:
+            parameters["strain_type"].SetString("strain_xx");
+            break;
+        case StrainType::STRAIN_YY:
+            parameters["strain_type"].SetString("strain_yy");
+            break;
+        case StrainType::STRAIN_ZZ:
+            parameters["strain_type"].SetString("strain_zz");
+            break;
+        case StrainType::STRAIN_XY:
+            parameters["strain_type"].SetString("strain_xy");
+            break;
+        case StrainType::STRAIN_XZ:
+            parameters["strain_type"].SetString("strain_xz");
+            break;
+        case StrainType::STRAIN_YZ:
+            parameters["strain_type"].SetString("strain_yz");
+            break;
+    };
+
+    return parameters;
+}
+
+double StrainSensor::CalculateValue(ModelPart& rModelPart)
+{
+    double directional_strain = 0.0;
+    if (rModelPart.HasElement(mElementId)) {
+        auto& r_element = rModelPart.GetElement(mElementId);
+
+        std::vector<Matrix> strains;
+        r_element.CalculateOnIntegrationPoints(mrStrainVariable, strains, rModelPart.GetProcessInfo());
+
+        for (const auto& strain : strains) {
+            directional_strain += *(strain.data().begin() + mStrainType);
+        }
+
+        directional_strain *= (this->GetWeight() / strains.size());
+    }
+
+    return rModelPart.GetCommunicator().GetDataCommunicator().SumAll(directional_strain);
+}
+
+void StrainSensor::CalculateGradient(
+    const Element& rAdjointElement,
+    const Matrix& rResidualGradient,
+    Vector& rResponseGradient,
+    const ProcessInfo& rProcessInfo)
+{
+    KRATOS_TRY
+
+    if (rResponseGradient.size() != rResidualGradient.size1()) {
+        rResponseGradient.resize(rResidualGradient.size1(), false);
+    }
+
+    rResponseGradient.clear();
+
+    if (rAdjointElement.Id() == mElementId) {
+        auto& r_element = const_cast<Element&>(rAdjointElement);
+        std::vector<Matrix> perturbed_strains, ref_strains;
+        r_element.CalculateOnIntegrationPoints(mrStrainVariable, ref_strains, rProcessInfo);
+
+        const double delta = rProcessInfo[PERTURBATION_SIZE];
+        auto& r_geometry = r_element.GetGeometry();
+
+        IndexType local_index = 0;
+        for (IndexType i_node = 0; i_node < r_geometry.size(); ++i_node) {
+            auto& r_node = r_geometry[i_node];
+
+            if (r_node.HasDofFor(ADJOINT_DISPLACEMENT_X)) {
+                rResponseGradient[local_index++] = this->CalculateStrainDirectionalSensitivity(
```

Comment:
it looks like you start assigning the value not from 0 but from 1 -> why do you skip  localindex = 0? or I miss smth?

Reply:
`local_index++` is a post increment, so I am assigning values from index zero.